### PR TITLE
posix: semaphore: Add unit tests for POSIX semaphore

### LINKED ISSUE #1748: Add some unit test for semaphores
Add some unit tests for implemented routines from semaphore.h
```
//extern int    sem_close(sem_t *);
extern int    sem_destroy(sem_t *);
extern int    sem_getvalue(sem_t *, int *);
extern int    sem_init(sem_t *, int, unsigned);
//extern sem_t *sem_open(const char *, int, ...);
extern int    sem_post(sem_t *);
extern int    sem_timedwait(sem_t *, const struct timespec *); //this
extern int    sem_trywait(sem_t *); //OK
//extern int    sem_unlink(const char *);
extern int    sem_wait(sem_t *);```

### PR DESCRIPTION
**Pull request by Inokinoki**
Resolve #1748 Add some unit tests for:

```c
extern int    sem_destroy(sem_t *);
extern int    sem_getvalue(sem_t *, int *);
extern int    sem_init(sem_t *, int, unsigned);
extern int    sem_post(sem_t *);
extern int    sem_timedwait(sem_t *, const struct timespec *);
extern int    sem_trywait(sem_t *);
extern int    sem_wait(sem_t *);
```

Implement
```c
extern int    sem_timedwait(sem_t *, const struct timespec *);
```

### CODE REVIEW

File: src/tests/posix/sem_test.c [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    clock_gettime(CLOCK_REALTIME, &ts);
+    ts.tv_nsec += 100000000;		// Wait 0.1s
+    test_assert_equal(sem_timedwait(&semaphore, &ts), -ETIMEDOUT);
+    clock_gettime(CLOCK_REALTIME, &now);
+    test_assert(ts_cmp(&ts, &now, <=));
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, 0);
+}
+
+static struct thread *sem_post_thread;
```
Comment (anton-bondarev) [inline src/tests/posix/sem_test.c]: Can you use pthreads instead?

File: src/tests/posix/sem_test.c [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+TEST_CASE("Test sem_init with shared=1, value=1") {
+    sem_t semaphore;
+
+    test_assert_equal(sem_init(&semaphore, 1, 1), ENOERR);
+}
+
+TEST_CASE("Test sem_destroy") {
+    sem_t semaphore;
+
+    // Up to now, sem_destroy does nothing but return ENOERR
```
Comment (anton-bondarev) [inline src/tests/posix/sem_test.c]: please don't use `//` use only `/* */`

File: src/tests/posix/sem_test.c [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    clock_gettime(CLOCK_REALTIME, &ts);
+    ts.tv_nsec += 100000000;		/* Wait 0.1s */
+    test_assert_equal(sem_timedwait(&semaphore, &ts), -ETIMEDOUT);
+    clock_gettime(CLOCK_REALTIME, &now);
+    test_assert(ts_cmp(&ts, &now, <=));
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, 0);
+}
+
+static pthread_t sem_post_thread;
```
Comment (anton-bondarev) [inline src/tests/posix/sem_test.c]: Why is it global?
As I can see it is used only in `TEST_CASE("Test sem_timedwait") `
You can declare it as a local variable

File: src/tests/posix/sem_test.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+static sem_t semaphore_test_timedwait;
```
Comment (anton-bondarev) [inline src/tests/posix/sem_test.c]: You can declare it as a local variable and pass pointer to it to `sem_post_run(void *arg) ` as arg through `pthread_create(&sem_post_thread, NULL, sem_post_run, &semaphore_test_timedwait)`

**Comment by Inokinoki** [comment]
> LGTM
> Thank you.
> Please rebase you branch under embox/master and squash commits

Done.

Should I include the tests in the other templates? Up to now, I only added the test into x86/qemu.

**Comment by anton-bondarev** [comment]
> Done.
Great. Thanks.

> Should I include the tests in the other templates? Up to now, I only added the test into x86/qemu.
Yes, you should. Please include it in to x86/test/unit by the separate commit and also remove it from x86/qemu (you can make commit with removing this line and then squish commits).

**Pull request merged**